lib/pc_sort.c

```diff
@@ -0,0 +1,267 @@
+/***********************************************************************
+* pc_sort.c
+*
+*  Pointclound patch sorting.
+*
+*  PgSQL Pointcloud is free and open source software provided
+*  by the Government of Canada
+*  Copyright (c) 2013 Natural Resources Canada
+*
+*  Author: M. Brédif
+*
+***********************************************************************/
+#include "pc_api_internal.h"
+#include <assert.h>
+#include <stdlib.h>
+
+
+/**
+ * Comparators
+ */
+
+int
+pc_compare_dim (const void *a, const void *b, void *arg)
+{
+    PCDIMENSION **dim = (PCDIMENSION **)arg;
+    uint32_t byteoffset     = dim[0]->byteoffset;
+    uint32_t interpretation = dim[0]->interpretation;
+    double da = pc_double_from_ptr(a+byteoffset,interpretation);
+    double db = pc_double_from_ptr(b+byteoffset,interpretation);
+    int cmp = ((da > db) - (da < db));
+    return ( cmp == 0 && dim[1]) ? pc_compare_dim(a,b,dim+1) : cmp;
+}
+
+int
+pc_compare_pcb (const void *a, const void *b, const void *arg)
+{
+    PCBYTES *pcb = (PCBYTES *)arg;
+    double da = pc_double_from_ptr(a,pcb->interpretation);
+    double db = pc_double_from_ptr(b,pcb->interpretation);
+    return ((da > db) - (da < db));
+}
+
+
+/**
+ * Sort
+ */
+
+PCPATCH_UNCOMPRESSED *
+pc_patch_uncompressed_sort(const PCPATCH_UNCOMPRESSED *pu, PCDIMENSION **dim)
+{
+    PCPATCH_UNCOMPRESSED *spu = pc_patch_uncompressed_make(pu->schema, pu->npoints);
+    
+    memcpy(spu->data, pu->data, pu->datasize);
+    spu->npoints = pu->npoints;
+    spu->bounds  = pu->bounds;
+    spu->stats   = pc_stats_clone(pu->stats);
+    
+    qsort_r(spu->data, spu->npoints, pu->schema->size, pc_compare_dim, dim);
+    
+    return spu;
+}
+
+PCDIMENSION **pc_schema_get_dimensions_by_name(const PCSCHEMA *schema, const char ** name, int ndims)
+{
+    PCDIMENSION **dim = pcalloc( (ndims+1) * sizeof(PCDIMENSION *));
+    int i;
+    for(i=0; i<ndims; ++i) 
+    {
+        dim[i] = pc_schema_get_dimension_by_name(schema, name[i]);
+        if ( ! dim[i] ) {
+            pcerror("dimension \"%s\" does not exist", name[i]);
+            return NULL;
+        }
+        assert(dim[i]->scale>0);
+    }
+    dim[ndims] = NULL;
```

Comment:
I see you are here terminating the list with a NULL. Maybe use a typedef to make this a PCDIMENSION_LIST so that it's clear that this list has a property of NULL termination.